### ISSUE TITLE
Inconsistent predict results

### ISSUE THREAD

**Issue by Mussil** [tensorflow/tensorflow#58997]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.5.0

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04.4 LTS

### Python version

3.8

### CUDA/cuDNN version

cuda-11.2

### GPU model and memory

GRID A100-20C, NVIDIA-SMI 510.47.03

### Current Behaviour?

```shell
Inconsistent predict result on Ubuntu 20.04.4 with A100 GPU when using Conv2D layer.
I'm running model.predict(X) few times with the identical loaded model from file and same input, but there are numeric difference in the results.

We performed multiple inferences (starting new process each time) with the following configuration:
• Windows 10 + CPU: same results each run 
• Ubuntu 18.04 + CPU: same results each run
• Ubuntu 18.04 + V100: same results each run 
• Ubuntu 20.04 + A100: different results each run (!!) 
```

### Standalone code to reproduce the issue

```shell
# Simple model that illustrate the problem:

# First- create the model and save it. ( Relayed on: https://iq.opengenus.org/conv2d-in-tf/)

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import MaxPooling2D, Flatten, Dense

fashion = keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion.load_data()
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255


model = Sequential()
model.add(Conv2D(32, (3, 3),input_shape=(28, 28, 1)))
# model.add(MaxPooling2D((2, 2)))
# model.add(Conv2D(64, (5, 5), activation='relu'))
model.add(Flatten())
model.add(Dense(10, activation='softmax'))
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('Test accuracy:', test_acc)


model.save('model.h5')


# -----------------------------------------------------------------------------
# Then: 
# load the model and run it few times:
model= tf.keras.models.load_model('model.h5')
print(model.predict(test_images))

# Gives different predictions


# You might need to remove cache files from one run to another to notice the differences: 
# sudo rm -rf ~/.nv/
```

### Relevant log output

```shell
# for example from first predict run:
[[1.9118926e-16 2.1664609e-21 1.9678958e-13 ... 3.2384528e-13
  8.9434872e-15 1.0000000e+00]
 [1.4733767e-13 6.6338912e-17 9.9999976e-01 ... 1.9979627e-28
  6.2790036e-15 9.4947700e-20]
 [3.4300227e-15 1.0000000e+00 1.4307653e-19 ... 1.5259588e-22
  4.3525246e-16 2.7868801e-20]
 ...
 [6.6930616e-19 1.8571309e-32 4.7568495e-30 ... 0.0000000e+00
  1.0000000e+00 2.0719716e-37]
 [5.6024744e-17 1.0000000e+00 1.7536839e-21 ... 2.5832318e-19
  6.5471020e-19 4.7896584e-16]
 [3.2346847e-10 1.4122374e-10 3.9109116e-07 ... 1.5287013e-05
  3.6158244e-08 1.4269066e-05]]

# 2nd predict run:
[[1.9118926e-16 2.1664446e-21 1.9678996e-13 ... 3.2384528e-13
  8.9434533e-15 1.0000000e+00]
 [1.4733681e-13 6.6340434e-17 9.9999976e-01 ... 1.9982675e-28
  6.2802491e-15 9.4948418e-20]
 [3.4302713e-15 1.0000000e+00 1.4307872e-19 ... 1.5260113e-22
  4.3524746e-16 2.7868801e-20]
 ...
 [6.6841317e-19 1.8562527e-32 4.7557244e-30 ... 0.0000000e+00
  1.0000000e+00 2.0703124e-37]
 [5.6203482e-17 1.0000000e+00 1.7566836e-21 ... 2.5900697e-19
  6.5718233e-19 4.8014216e-16]
 [3.2395439e-10 1.4134931e-10 3.9128366e-07 ... 1.5299629e-05
  3.6214352e-08 1.4277887e-05]]

# (can be noticed on the last value)
```
</details>

**Comment by tiruk007** [comment]
@Mussil 
I was able to get the same predictions from the given code on colab using TF v2.11. Could you please try with the latest version of tensorflow which is TF v2.11 and find gist of [cpu](https://colab.research.google.com/gist/tiruk007/bc6d8f563161d81550379bab07a3aa51/58997cpu.ipynb) & [gpu](https://colab.research.google.com/gist/tiruk007/ee6739fcbdbffc42476c77301eaf9511/58997gpu.ipynb) for reference.

Thank you!

**Comment by Mussil** [comment]
I'm not sure which GPU is used in [gpu](https://colab.research.google.com/gist/tiruk007/ee6739fcbdbffc42476c77301eaf9511/58997gpu.ipynb) , I faced this issue on A100 GPU in particular. 

Anyway upgrade to TF v2.11 didn't help.

**Comment by SuryanarayanaY** [comment]
Hi @Mussil ,
I have tried to replicate the issue you mentioned in a VM with Ubuntu 20.04, 2 x NVIDIA A100 40GB GPUs, x86/64 Architecture and with TF2.10 but got same results in both runs(predictions).Please refer to attached results.
<img width="1512" alt="Screenshot 2022-12-27 at 2 47 10 PM" src="https://user-images.githubusercontent.com/116063290/209644056-ba997b60-6a9a-40b5-8bc9-31b6edb6fa15.png">

Please refer to attached output of `nvidia-smi`
<img width="1507" alt="Screenshot 2022-12-27 at 2 46 06 PM" src="https://user-images.githubusercontent.com/116063290/209644149-deee1c67-961e-479b-9ac9-52bf7e2e765a.png">
Please refer attached output of GPU list

<img width="1503" alt="Screenshot 2022-12-27 at 2 54 25 PM" src="https://user-images.githubusercontent.com/116063290/209644690-8a5befb9-7854-4c2f-9cca-11493b6ea797.png">

Anyways TF2.5V is older versions and we recommend to use latest versions only.Can you test with TF 2.10V and recommended nvidia driver from my attached configuration?

Thankyou!

**Comment by Mussil** [comment]
Just to clarify did you try to load the model and predict in 2 different process and release cache?

**Comment by SuryanarayanaY** [comment]
Hi @Mussil ,
I used the below code for clearing cache.Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/26d9620f2329813e2e21eb6e6f5abb73/58997_vm2.ipynb) used i used in VM.
```
print(model.predict(test_images))
import subprocess
subprocess.run('sudo rm -rf ~/.nv/',shell=True)
print(model.predict(test_images))
```
To Cross check i used `tf.device('GPU')` for `model.predict()` and the results are same for both runs in [Colab](https://colab.research.google.com/gist/SuryanarayanaY/c249d353176be70c7bdffc53804d53bb/58997_colab.ipynb) and the VM as well.
<img width="1506" alt="Screenshot 2022-12-28 at 6 24 30 PM" src="https://user-images.githubusercontent.com/116063290/209816167-dbf6a08d-54bf-47d7-aee9-18cd808fca36.png">

**Comment by xmajzel** [comment]
Hello,

I noticed similar problem with Google Colab and own desktop computer in prediction difference. In my case the problem was that I trained model in Google Colab and saved model in h5 format to Google Drive. After loading model on localhost machine the prediction metrics were significantly different (more likely the weights were initialized randomly).

None of changing tensorflow and numpy random seed, manual variable initialization or setting learning phase did helped, but i noticed this in console every time I was running prediction (the warning/error did not stop code execution) on own desktop:
```bash
tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2, command: '"ptxas.exe" "--version"'
```

Running `conda install -c nvidia cuda-nvcc` in my case solved the problem (I found solution here: https://stackoverflow.com/questions/66946774/call-to-createprocess-failed-error-code-2-tensorflow).
